modules/flowable-cmmn-engine/src/main/java/org/flowable/cmmn/engine/impl/delete/ComputeDeleteHistoricCaseInstanceIdsJobHandler.java

```diff
@@ -0,0 +1,574 @@
+            boolean startDeletion = !DeleteCaseInstanceBatchConstants.STATUS_FAILED.equals(batch.getStatus());
+
+            for (BatchPart nextComputePart : nextComputeParts) {
+                if (!nextComputePart.getId().equals(batchPart.getId())) {
+                    startDeletion = false;
+                    JobService jobService = engineConfiguration.getJobServiceConfiguration().getJobService();
+
+                    JobEntity nextComputeJob = jobService.createJob();
+                    nextComputeJob.setJobHandlerType(ComputeDeleteHistoricCaseInstanceIdsJobHandler.TYPE);
+                    nextComputeJob.setJobHandlerConfiguration(nextComputePart.getId());
+                    nextComputeJob.setScopeType(ScopeTypes.CMMN);
+                    jobService.createAsyncJob(nextComputeJob, false);
+                    jobService.scheduleAsyncJob(nextComputeJob);
+                    break;
+                }
+            }
+
+            if (startDeletion) {
+                JobService jobService = engineConfiguration.getJobServiceConfiguration().getJobService();
+                JobEntity nextDeleteJob = jobService.createJob();
+                nextDeleteJob.setJobHandlerType(DeleteHistoricCaseInstanceIdsJobHandler.TYPE);
+                nextDeleteJob.setJobHandlerConfiguration(batchPartForDelete.getId());
+                nextDeleteJob.setScopeType(ScopeTypes.CMMN);
+                jobService.createAsyncJob(nextDeleteJob, false);
+                jobService.scheduleAsyncJob(nextDeleteJob);
+            }
+        }
+
+        batchService.completeBatchPart(batchPart.getId(), DeleteCaseInstanceBatchConstants.STATUS_COMPLETED, resultNode.toString());
+    }
+
+    private void failBatchPart(CmmnEngineConfiguration engineConfiguration, BatchService batchService, BatchPart batchPart, Batch batch,
+            String resultJson, boolean sequentialExecution) {
+        batchService.completeBatchPart(batchPart.getId(), DeleteCaseInstanceBatchConstants.STATUS_FAILED, resultJson);
+        if (sequentialExecution) {
+            completeBatch(batch, DeleteCaseInstanceBatchConstants.STATUS_FAILED, engineConfiguration);
+        }
+    }
+
+    protected void completeBatch(Batch batch, String status, CmmnEngineConfiguration engineConfiguration) {
+        engineConfiguration.getBatchServiceConfiguration()
+                .getBatchService()
+                .completeBatch(batch.getId(), status);
+    }
+
+    protected HistoricCaseInstanceQuery createQuery(JsonNode queryNode, CmmnEngineConfiguration engineConfiguration) {
+        HistoricCaseInstanceQuery query = engineConfiguration.getCmmnHistoryService()
+                .createHistoricCaseInstanceQuery();
+
+        populateQuery(queryNode, query, engineConfiguration);
+        return query;
+    }
+
+    protected void populateQuery(JsonNode queryNode, HistoricCaseInstanceQuery query, CmmnEngineConfiguration engineConfiguration) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 133 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.